[detector] feat: added rootly detector

sahil9001 opened the pull request:
<!--
Please create an issue to collect feedback prior to feature additions. Please also reference that issue in any PRs.
If possible try to keep PRs scoped to one feature, and add tests for new features.
-->

### Description:
Explain the purpose of the PR.
Relates to #3396 
I have verified the tests for rootly by trying it with original working API key and here are the results.

<img width="750" alt="Screenshot 2024-10-16 at 8 33 32 PM" src="https://github.com/user-attachments/assets/9cd4a4c9-da6f-41ad-96a0-a0bbd3150fb7">

Detector Unit Tests:
![image](https://github.com/user-attachments/assets/4c6befa0-1113-4f75-8b6c-a1d55c5d1ea4)


### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?

sahil9001 committed:
[detector] feat: added rootly detector

rgmz commented on pkg/detectors/rootly/rootly.go:
```diff
@@ -0,0 +1,77 @@
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"rootly"}
```
```suggestion
	return []string{"rootly_"}
```

rgmz commented on pkg/detectors/rootly/rootly.go:
```diff
@@ -0,0 +1,77 @@
+	matches := keyPat.FindAllStringSubmatch(dataStr, -1)
+
+	for _, match := range matches {
+		if len(match) != 1 {
+			continue
+		}
+		resMatch := strings.TrimSpace(match[0])
```
Just curious, how did you generate this? This logic, present in some detectors, is old and doesn't conform to current best practices.

The "alchemy" detector seems to be the template that is good to follow.
https://github.com/trufflesecurity/trufflehog/blob/main/hack/docs/Adding_Detectors_external.md
https://github.com/trufflesecurity/trufflehog/blob/aa17b8eba486bbdf19bc77aedb3c56f9cb00e1ac/pkg/detectors/alchemy/alchemy.go#L4

sahil9001 replied:
I actually took the reference from `alienvault` and tried to follow that, I will update this.

sahil9001 replied:
@rgmz Can you check again?

sahil9001 replied:
<img width="369" alt="Screenshot 2024-10-16 at 12 25 23 PM" src="https://github.com/user-attachments/assets/d1af7d6c-e34b-4605-bfb2-17a5874cc8d0">

sahil9001 committed:
[wip]feat: detector rootly

sahil9001 committed:
feat: added correct rootly detector

sahil9001 committed:
feat: added correct unit tests

sahil9001 committed:
feat: added correct unit tests

sahil9001 committed:
feat: added correct unit tests

sahil9001 committed:
feat: added correct unit test

kashifkhan0771 commented:
@sahil9001 Can you please resolve conflicts on this PR?

kashifkhan0771 commented:
I am closing [this](https://github.com/trufflesecurity/trufflehog/pull/3418) duplicate PR.

nabeelalam commented on pkg/detectors/rootly/rootly.go:
```diff
@@ -0,0 +1,77 @@
+		}
+
+		if verify {
+			req, err := http.NewRequestWithContext(ctx, "GET", "https://api.rootly.com/v1/users/me", nil)
+			if err != nil {
+				continue
+			}
+			req.Header.Add("Authorization", "Bearer "+resMatch)
+			res, err := client.Do(req)
+			if err == nil {
+				defer res.Body.Close()
+				if res.StatusCode >= 200 && res.StatusCode < 300 {
+					s1.Verified = true
+				}
+			}
```
Hey @sahil9001, thanks for the pull-request. As referenced in an older comment, the [Alchemy detector](https://github.com/trufflesecurity/trufflehog/blob/aa17b8eba486bbdf19bc77aedb3c56f9cb00e1ac/pkg/detectors/alchemy/alchemy.go) is a good reference for adding new detectors.
The way we handle verification includes setting a `VerificationError` in the results object in cases where the response is unexpected.

sahil9001 committed:
feat: added `VerificationError` abstraction

Signed-off-by: Sahil Silare <sahilsilare@gmail.com>

nabeelalam commented on pkg/detectors/rootly/rootly.go:
```diff
@@ -0,0 +1,90 @@
+    return results, nil
+}
+
+func verifyMatch(ctx context.Context, client *http.Client, token string) (bool, map[string]string, error) {
```
nit: Since we're not returning any extra data in the `verifyMatch`, `map[string]string` doesn't have to be returned, and we don't need the line `s1.ExtraData = extraData` either.

sahil9001 committed:
feat: addressed review comments

Signed-off-by: Sahil Silare <sahilsilare@gmail.com>

amanfcp committed:
Merge branch 'main' of github.com:trufflesecurity/trufflehog into detector-rootly

amanfcp committed:
Changed verification URL
modified status code handling and added comments
fixed GCP cred access in integration tests

amanfcp commented:
<img width="769" height="371" alt="Screenshot from 2025-07-29 16-40-24" src="https://github.com/user-attachments/assets/4a441be2-7f35-44cd-af61-807248aa3d15" />

amanfcp committed:
Refactor HTTP request in Rootly detector to use http.MethodGet and http.NoBody for improved clarity and consistency.

shahzadhaider1 commented on pkg/detectors/rootly/rootly.go:
```diff
@@ -0,0 +1,95 @@
+	client = common.SaneHttpClient()
+
+	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
+	keyPat = regexp.MustCompile(`\brootly_[a-f0-9]{64}\b`)
```
This looks good, but it might be more robust if we wrap the regex in a capturing group

shahzadhaider1 reviewed (approved):
Looks good. Few minor comments.

shahzadhaider1 committed:
Merge branch 'main' into detector-rootly

amanfcp committed:
Incorporated feedback

amanfcp committed:
Merge branch 'detector-rootly' of https://github.com/sahil9001/trufflehog into detector-rootly

amanfcp committed:
Merge branch 'main' of github.com:trufflesecurity/trufflehog into detector-rootly

amanfcp committed:
fixed missing bracket

shahzadhaider1 committed:
Merge branch 'main' into detector-rootly
